### ISSUE TITLE
Add Vault implementation details for the Frontiers Vault paper submission

### ISSUE THREAD

**Issue by praeducer** [MRN-Code/coinstac#1631]
We want to discuss the specific technologies involved and how they are configured. Linking to actual code and documenting how the code works.

Use case: Another open source/science software development team could read the paper and know how to implement a Vaults type feature in their system. Say there is some COINSTAC that specializes in hearts instead of brains and they needed to do something like this. 

Another perspective is to write up something that is detailed enough that if we didn't already implement Vaults we would have the details we needed to implement it.

**Comment by praeducer** [comment]
Vince says this would be a good value add. It's not necessary that there is enough detail to rebuild this feature for submission, but it fits the spirit of reproducibility and open science. This is especially reasonable if we can do it in a single paragraph.

**Comment by praeducer** [comment]
- [x] Mention Docker
- [x] Mention Vaults are abstractly equivalent to users
- [x] Link to code and getting started documentation

**Comment by praeducer** [comment]
We may be able to borrow some vocabulary and thoughts on how Vaults integrate with the greater ecosystem here: https://oreoni.github.io/04/data.html

**Comment by praeducer** [comment]
Why are doing this?
- Show how it leverages existing patterns in the application. KISS. Advantages of this are leverage things like UI components and how user security is managed. 
- Engineering readers may be curious and want to know how it is built and how works. Teach them both architectural patterns and software development principles.
- Attract and ramp up contributors.
- Get feedback on our designs for future work.
- 

Ross feels like it is kind the same old stuff we've been doing. It is an extension of how the normal client will work.

Works how the user workflow works. Vault client consumes data in a more automated and static way. It wasn't a huge jump to implement. Leverages the current system pretty much entirely. One dev was able to implement it. 400-500 lines of code and some adjustments. 

How does the current user workflow/client work? Receives notifications with meta data on how to run. Pipeline server receives API call on internal server from graphql to start a pipeline. Then the remote node starts first, then the clients get notifications to join. Through graphql subscriptions, which is a web socket. Using Apollo for client side JavaScript. We can have several Vault clients running together. They are kind of collaborative like users would be. Challenge is normalizing and relating the data together.

How it is different? Uploads data to S3. Vault client has a small percentage of GUI client. Instead of having an API server listening for a call to start a pipeline, it is listening on a subscription. It is more like a server sitting to waiting on data. It is completely decoupled from the GUI. Vault node can be several clients at once. Big difference is how data is required. Users need to choose from static data set. This is displayed in the UI so folks can choose Vaults. A challenge here is have the Vault data map itself to arbitrary pipelines. Normally real world users can map data in a way that makes sense. Normally we choose how to map as users, with Vaults we do this ahead of time. Normally users decide what their columns names are and these are stored as a template in a database (MongoDB). When a run happens, column names get mapped. For a Vault client, since it can't map things, we store along with the Vault profile we store the CSV columns. This is stored in a different MongoDB collection than normal users. The schemas are different. When we create a pipeline with a Vault user, it is repopulated with all of the data stored in the Vault database.

All of our Vaults are on one server. In the future, could put it out in Kubernetes. This would make sense if we had more load, like number of computations running that hog CPU bandwidth or RAM. Offloading the docker/container system would be most of the load. It works like a normal client does. There is a container for every run of a Vault client. What is a run? A run is an instance of a pipeline on a consortia. A consortia has many runs and many pipelines. Runs are a real world instantiation of pipelines. When we click start, that is a run. Graphql is listening for new runs. That is the change document is it looking for. At the moment, we only have one active run per consortia. But many consortia can leverage a Vault. We say the Vault user is the template we are mapping to. 

Main use case for Vince is for people to be able to run analyses with resources or data they have, without necessarily having their own data. "You can do research with research data without computational knowledge or data of your own." A primary contribution to science is decreasing costs of data access. Costs are time, money, study participants, and machinery. Accessibility to well known computations and useful data sets. It can provide data that you may never usually be able to operate on. COINSTAC can help you know what computations or data to use. Think more about the many to many relationships of data and computations. Talk about data accessibility and liquidity (can put it to work fast).

Future work? Resource management of 1. devops / hardware specs / load balancing 2. front-end API perspective like by managing runs, instances, or connections. Currently we manually map data for Vaults. In the future could normalize the data into a big transfer table (this is something missing in our field in general!). Or put it on the users uploading Vaults to follow standards more closely. Could be standards for imaging data. Standardizing in how you actually process things can accentuate things that are worthy of analysis. 

What is the name of the S3 buckets and server for me to examine? 

What is the link to the Vault implementation? Note: watch out for confusing structures. File called subscriptions is where most things happen. Consumes and processes graphql messages. 

What existing COINSTAC packages are we leverage and how are they used?

Anything other context I may need to know before reviewing the code and cloud resources?

**Comment by praeducer** [comment]
https://github.com/trendscenter/coinstac/blob/master/packages/coinstac-vault-client/src/subscription.js

**Comment by praeducer** [comment]
https://s3.console.aws.amazon.com/s3/buckets/coinstac-run-assets?region=us-east-1&tab=objects
https://s3.console.aws.amazon.com/s3/buckets/coinstac-run-assets-dev?region=us-east-2&tab=objects

**Comment by praeducer** [comment]
Other approaches:
- Add to supplementary section
- Link to Github code or documentation/wiki

**Comment by praeducer** [comment]
What goes in the methods section?

- Vault system
     - How it is built
     - How to use it
          - as a user
          - as a data owner
     - How it is helpful
- Curated vaults
     - What are they
     - How they are helpful

**Comment by praeducer** [comment]
Add in Github data.
https://github.com/trendscenter/coinstac



![Image](https://user-images.githubusercontent.com/4081511/221015032-26224ee4-dbaf-47d1-8925-6f9166fe71e9.png)

**Comment by praeducer** [comment]
Need to build off of the COINSTAC and Vaults Use Cases sections. See related task: https://github.com/trendscenter/coinstac/issues/1667

**Comment by praeducer** [comment]
![Image](https://user-images.githubusercontent.com/4081511/221017617-035ecfbb-46e5-4f39-9f8a-7a4f97097bd8.png)
